File: src/msra_lsp/analysis.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable

from .model import (
    AssignmentDef,
    Diagnostic,
    Expr,
    IdentExpr,
    InlineTableExpr,
    ParsedDocument,
    Position,
    Range,
    RefExpr,
    ReferenceOccurrence,
    SequenceExpr,
    TableDef,
)


KNOWN_DYNAMIC_ROOTS = {
    "UNSTANDART_HEADERS",
    "CAPTURED_URLS",
    "COOKIES",
    "LOCAL_STORAGE",
    "SESSION_STORAGE",
}

VIRTUAL_ROOTS = {
    "DOCUMENT",
    "VARIABLES",
    "INPUT",
}


@dataclass(slots=True)
class AnalysisResult:
    document: ParsedDocument
    diagnostics: list[Diagnostic] = field(default_factory=list)
    table_index: dict[tuple[str, ...], TableDef] = field(default_factory=dict)
    assignment_index: dict[tuple[str, ...], AssignmentDef] = field(default_factory=dict)
    root_table_index: dict[str, tuple[str, ...]] = field(default_factory=dict)


def analyze_document(document: ParsedDocument) -> AnalysisResult:
    diagnostics = list(document.diagnostics)
    table_index = dict(document.tables)
    assignment_index = dict(document.assignments)
    root_table_index = {
        ".".join(path): path
        for path in table_index
        if path
    }
    result = AnalysisResult(
        document=document,
        diagnostics=diagnostics,
        table_index=table_index,
        assignment_index=assignment_index,
        root_table_index=root_table_index,
    )

    for ref in document.references:
        resolved = resolve_reference(ref, result)
        ref.resolved_path = resolved[0] if resolved else None
        ref.resolved_kind = resolved[1] if resolved else None
        if resolved is None and ref.expr.parts:
            root = ref.expr.parts[0].value
            if root not in KNOWN_DYNAMIC_ROOTS and root not in VIRTUAL_ROOTS:
                diagnostics.append(
                    Diagnostic(
                        f"Unresolved reference <{render_ref(ref.expr)}>",
                        ref.range,
                        code="unresolved-reference",
                    )
                )
    return result


def render_ref(ref: RefExpr) -> str:
    rendered: list[str] = []
    for part in ref.parts:
        if part.kind == "name":
            if rendered:
                rendered.append(".")
            rendered.append(str(part.value))
        elif part.kind == "index":
            rendered.append("[...]")
        elif part.kind == "call":
            rendered.append("(...)")
    return "".join(rendered)


def resolve_reference(ref: ReferenceOccurrence, result: AnalysisResult) -> tuple[tuple[str, ...], str] | None:
    path = ref_path_segments(ref.expr)
    if not path:
        return None
    table_path = ref.table_path
    expanded = expand_virtual_path(path, table_path)
    if expanded is not None:
        resolved = resolve_static_path(expanded, result)
        if resolved is not None:
            return resolved
    resolved = resolve_static_path(tuple(path), result)
    if resolved is not None:
        return resolved
    # fallback for virtual roots that have no static target
    root = path[0]
    if root in KNOWN_DYNAMIC_ROOTS:
        return tuple(path), "dynamic-root"
    return None


def ref_path_segments(ref: RefExpr) -> list[str]:
    path: list[str] = []
    for part in ref.parts:
        if part.kind != "name":
            break
        path.append(str(part.value))
    return path


def expand_virtual_path(path: list[str], current_table: tuple[str, ...]) -> tuple[str, ...] | None:
    if not path:
        return None
    root = path[0]
    if root == "DOCUMENT":
        if len(path) >= 2 and path[1] == "PREFIXES":
            return ("app", "prefixes", *path[2:])
        if len(path) >= 2 and path[1] in {"REGEX", "REGEXES"}:
            return ("app", "regexes", *path[2:])
        if len(path) >= 2 and path[1] == "WARMUP":
            return ("app", "warmup", *path[2:])
    if root == "VARIABLES":
        return ("app", "variables", *path[1:])
    if root == "INPUT":
        function_id = current_function_id(current_table)
        if function_id:
            return ("app", "func", function_id, "input", *path[1:])
    return None


def current_function_id(table_path: tuple[str, ...]) -> str | None:
    segments = list(table_path)
    for index in range(len(segments) - 2):
        if segments[index] == "app" and segments[index + 1] == "func":
            return segments[index + 2]
    return None


def resolve_static_path(path: tuple[str, ...], result: AnalysisResult) -> tuple[tuple[str, ...], str] | None:
    if path in result.assignment_index:
        return path, "assignment"
    if path in result.table_index:
        return path, "table"
    if not path:
        return None
    # Allow direct table access to nested symbols by matching prefixes.
    if len(path) > 1:
        for index in range(len(path) - 1, 0, -1):
            prefix = path[:index]
            suffix = path[index:]
            if prefix in result.table_index and suffix and len(suffix) == 1:
                candidate = prefix + suffix
                if candidate in result.assignment_index:
                    return candidate, "assignment"
    return None


def collect_definition_locations(result: AnalysisResult) -> dict[tuple[str, ...], Range]:
    locations: dict[tuple[str, ...], Range] = {}
    for path, table in result.table_index.items():
        locations[path] = table.header_range
    for path, assignment in result.assignment_index.items():
        locations[path] = assignment.key_range
    return locations


def table_children(result: AnalysisResult, table_path: tuple[str, ...]) -> list[tuple[str, ...]]:
    children: list[tuple[str, ...]] = []
    for path in result.table_index:
        if len(path) == len(table_path) + 1 and path[: len(table_path)] == table_path:
            children.append(path)
    return sorted(children)


def iter_all_paths(result: AnalysisResult) -> Iterable[tuple[str, ...]]:
    yield from sorted(result.table_index)
    yield from sorted(result.assignment_index)

File: vscode-extension/lsp/analysis.js
const {
  Diagnostic,
  comparePaths,
  pathLabel,
} = require("./model");
const { validateAssignmentRelations } = require("./assignment-relations");
const { validateAssignment } = require("./assignment-schema");
const { validateGroupAssignments } = require("./group-relations");
const { validateTableRelations } = require("./table-relations");
const { validateTablePath } = require("./path-schema");
const { pathIdentityKey, normalizePathSegments } = require("./path-schema");
const { parseFuncResultReference } = require("./reference-context");

const KNOWN_DYNAMIC_ROOTS = new Set([
  "UNSTANDARD_HEADERS",
  "CAPTURED_URLS",
  "COOKIES",
  "LOCAL_STORAGE",
  "SESSION_STORAGE",
]);

class AnalysisResult {
  constructor(document) {
    this.document = document;
    this.diagnostics = [];
    this.tableIndex = new Map();
    this.assignmentIndex = new Map();
    this.rootTableIndex = new Map();
  }
}

function analyzeDocument(document, context = {}) {
  const diagnostics = [...document.diagnostics];
  diagnostics.push(...validateMsrafStructure(context.rawDocument || null, context.sourcePath || document.uri || ""));
  const localTableIndex = new Map(document.tables);
  const localAssignmentIndex = new Map(document.assignments);
  const lookupTableIndex = context.tableIndex ? new Map([...context.tableIndex, ...localTableIndex]) : localTableIndex;
  const lookupAssignmentIndex = context.assignmentIndex ? new Map([...context.assignmentIndex, ...localAssignmentIndex]) : localAssignmentIndex;
  const rootTableIndex = new Map();
  for (const table of localTableIndex.values()) {
    const tablePath = table.pathSegments || table.path;
    if (!tablePath || !tablePath.length) {
      continue;
    }
    const validation = validateTablePath(tablePath);
    if (!validation.valid) {
      diagnostics.push(
        new Diagnostic(
          validation.message,
          tablePathSegmentRange(table, validation.segmentIndex),
          "error",
          "msra",
          validation.code,
        ),
      );
    }
  }
  for (const table of lookupTableIndex.values()) {
    const tablePath = table.pathSegments || table.path;
    if (tablePath && tablePath.length) {
      rootTableIndex.set(pathLabel(tablePath), table.path);
    }
  }
  diagnostics.push(...validateTableRelations(localTableIndex, lookupTableIndex));
  for (const assignment of localAssignmentIndex.values()) {
    const table = localTableIndex.get(pathIdentityKey(assignment.tablePath));
    for (const diagnostic of validateAssignment(assignment.tablePath, assignment, table ? table.assignments || [] : [])) {
      diagnostics.push(diagnostic);
    }
  }
  diagnostics.push(...validateAssignmentRelations(localTableIndex, localAssignmentIndex, lookupTableIndex, lookupAssignmentIndex));
  diagnostics.push(...validateGroupAssignments(lookupTableIndex, localAssignmentIndex));
  const result = new AnalysisResult(document);
  result.diagnostics = diagnostics;
  result.tableIndex = lookupTableIndex;
  result.assignmentIndex = lookupAssignmentIndex;
  result.rootTableIndex = rootTableIndex;

  for (const ref of document.references) {
    const funcResultReference = parseFuncResultReference(ref.expr, ref.tablePath || [], ref.valuePathSegments || []);
    if (funcResultReference !== null) {
      if (!funcResultReference.valid) {
        diagnostics.push(
          new Diagnostic(
            funcResultReference.message,
            funcResultReference.range || ref.range,
            "error",
            "msra",
            funcResultReference.code,
          ),
        );
        continue;
      }
      const resolved = resolveFuncResultReference(funcResultReference, result);
      if (resolved !== null) {
        ref.resolvedPath = resolved.path;
        ref.resolvedPathSegments = resolved.pathSegments;
        ref.resolvedPathKey = resolved.key;
        ref.resolvedKind = resolved.kind;
        continue;
      }
      diagnostics.push(
        new Diagnostic(
          `Unresolved reference <${renderRef(ref.expr)}>`,
          ref.range,
          "error",
          "msra",
          "unresolved-reference",
        ),
      );
      continue;
    }
    const resolved = resolveReference(ref, result);
    ref.resolvedPath = resolved ? resolved.path : null;
    ref.resolvedPathSegments = resolved ? resolved.pathSegments : null;
    ref.resolvedPathKey = resolved ? resolved.key : null;
    ref.resolvedKind = resolved ? resolved.kind : null;
    if (resolved === null && ref.expr.parts.length) {
      const root = ref.expr.parts[0].value;
      if (!KNOWN_DYNAMIC_ROOTS.has(root)) {
        diagnostics.push(
          new Diagnostic(
          `Unresolved reference <${renderRef(ref.expr)}>`,
            ref.range,
            "error",
            "msra",
            "unresolved-reference",
          ),
        );
      }
    }
  }

  return result;
}

function validateMsrafStructure(rawDocument, sourcePath) {
  if (!rawDocument || !String(sourcePath || "").toLowerCase().endsWith(".msraf")) {
    return [];
  }
  const diagnostics = [];
  for (const table of rawDocument.tables.values()) {
    const segments = table.pathSegments || table.path || [];
    const prefixIndex = findAppFuncPrefixIndex(segments);
    if (prefixIndex === -1) {
      continue;
    }
    diagnostics.push(
      new Diagnostic(
        'In .msraf files, omit the leading "app.func" namespace and write relative function tables instead.',
        segments[prefixIndex] && segments[prefixIndex].range ? segments[prefixIndex].range : table.headerRange,
        "error",
        "msra",
        "invalid-msraf-table-path",
      ),
    );
  }
  return diagnostics;
}

function findAppFuncPrefixIndex(segments) {
  for (let index = 0; index < segments.length - 1; index += 1) {
    const current = segments[index];
    const next = segments[index + 1];
    if (
      current &&
      current.value === "app" &&
      !current.quoted &&
      next &&
      next.value === "func" &&
      !next.quoted
    ) {
      return index;
    }
  }
  return -1;
}

function tablePathSegmentRange(table, segmentIndex) {
  if (table.pathSegments && table.pathSegments[segmentIndex] && table.pathSegments[segmentIndex].range) {
    return table.pathSegments[segmentIndex].range;
  }
  return table.headerRange;
}

function renderRef(ref) {
  const rendered = [];
  for (const part of ref.parts) {
    if (part.kind === "name") {
      if (rendered.length) {
        rendered.push(".");
      }
      rendered.push(String(part.value));
    } else if (part.kind === "index") {
      rendered.push("[...]");
    } else if (part.kind === "call") {
      rendered.push("(...)");
    }
  }
  return rendered.join("");
}

function resolveReference(ref, result) {
  const path = refPathSegments(ref.expr);
  if (!path.length) {
    return null;
  }
  const tablePath = ref.tablePath;
  const expanded = expandVirtualPath(path, tablePath);
  if (expanded !== null) {
    const resolved = resolveStaticPath(expanded, result);
    if (resolved !== null) {
      return resolved;
    }
  }
  const resolved = resolveStaticPath(path, result);
  if (resolved !== null) {
    return resolved;
  }
  const root = path[0] && path[0].value;
  if (KNOWN_DYNAMIC_ROOTS.has(root)) {
    return {
      path: path.map((segment) => String(segment && segment.value !== undefined ? segment.value : segment)),
      pathSegments: path,
      key: pathIdentityKey(path),
      kind: "dynamic-root",
    };
  }
  return null;
}

function resolveFuncResultReference(funcResultReference, result) {
  const functionPath = [makePathSegment("app"), makePathSegment("func"), makePathSegment(funcResultReference.functionId)];
  const resolved = resolveStaticPath(functionPath, result);
  if (resolved === null) {
    return null;
  }
  return {
    path: resolved.path,
    pathSegments: resolved.pathSegments,
    key: resolved.key,
    kind: `func-result-${String(funcResultReference.resultKind || "").toLowerCase()}`,
  };
}

function refPathSegments(ref) {
  const path = [];
  for (const part of ref.parts) {
    if (part.kind !== "name") {
      break;
    }
    path.push({
      value: String(part.value),
      quoted: Boolean(part.quoted),
      range: part.range || null,
    });
  }
  return path;
}

function expandVirtualPath(path, currentTable) {
  if (!path.length) {
    return null;
  }
  const root = path[0] && path[0].value;
  if (root === "DOCUMENT") {
    if (path.length >= 2 && path[1] && path[1].value === "PREFIXES") {
      return [makePathSegment("app"), makePathSegment("prefixes"), ...path.slice(2)];
    }
    if (path.length >= 2 && path[1] && path[1].value === "REGEXES") {
      return [makePathSegment("app"), makePathSegment("regexes"), ...path.slice(2)];
    }
    if (path.length >= 2 && path[1] && path[1].value === "WARMUP") {
      return [makePathSegment("app"), makePathSegment("warmup"), ...path.slice(2)];
    }
  }
  if (root === "VARIABLES") {
    return [makePathSegment("app"), makePathSegment("variables"), ...path.slice(1)];
  }
  if (root === "GROUPS") {
    return [makePathSegment("app"), makePathSegment("groups"), ...path.slice(1)];
  }
  if (root === "INPUT") {
    const functionId = currentFunctionId(currentTable);
    if (functionId) {
      return [
        makePathSegment("app"),
        makePathSegment("func"),
        makePathSegment(functionId),
        makePathSegment("input"),
        ...path.slice(1),
      ];
    }
  }
  return null;
}

function currentFunctionId(tablePath) {
  const segments = [...tablePath];
  for (let index = 0; index < segments.length - 2; index += 1) {
    if (segments[index] === "app" && segments[index + 1] === "func") {
      return segments[index + 2];
    }
  }
  return null;
}

function resolveStaticPath(path, result) {
  if (path.length > 1) {
    const prefix = path.slice(0, -1);
    const prefixKey = pathIdentityKey(prefix);
    if (result.tableIndex.has(prefixKey)) {
      const last = path[path.length - 1];
      const candidateKey = assignmentIdentityKey(prefixKey, last && last.value);
      if (result.assignmentIndex.has(candidateKey)) {
        const assignment = result.assignmentIndex.get(candidateKey);
        return buildResolvedAssignment(prefix, last, assignment, candidateKey);
      }
    }
  }
  const key = pathIdentityKey(path);
  if (result.tableIndex.has(key)) {
    const table = result.tableIndex.get(key);
    return buildResolvedTable(table, key);
  }
  return null;
}

function collectDefinitionLocations(result) {
  const locations = new Map();
  for (const [key, table] of result.tableIndex.entries()) {
    locations.set(key, {
      sourcePath: table.sourcePath || null,
      range: table.headerRange,
    });
  }
  for (const [key, assignment] of result.assignmentIndex.entries()) {
    locations.set(key, {
      sourcePath: assignment.sourcePath || null,
      range: assignment.keyRange,
    });
  }
  return locations;
}

function tableChildren(result, tablePath) {
  const children = [];
  const parentSegments = normalizePathSegments(tablePath);
  const parentKey = pathIdentityKey(parentSegments);
  for (const table of result.tableIndex.values()) {
    const childSegments = table.pathSegments || normalizePathSegments(table.path);
    if (childSegments.length === parentSegments.length + 1 && pathIdentityKey(childSegments.slice(0, parentSegments.length)) === parentKey) {
      children.push(table.path);
    }
  }
  return children.sort(comparePaths);
}

function* iterAllPaths(result) {
  const tablePaths = [...result.tableIndex.values()].map((table) => table.path).sort(comparePaths);
  const assignmentPaths = [...result.assignmentIndex.values()].map((assignment) => assignment.fullPath).sort(comparePaths);
  yield* tablePaths;
  yield* assignmentPaths;
}

module.exports = {
  AnalysisResult,
  KNOWN_DYNAMIC_ROOTS,
  analyzeDocument,
  collectDefinitionLocations,
  currentFunctionId,
  expandVirtualPath,
  iterAllPaths,
  refPathSegments,
  renderRef,
  resolveReference,
  tableChildren,
};

function makePathSegment(value, quoted = false) {
  return {
    value: String(value),
    quoted: Boolean(quoted),
    range: null,
  };
}

function assignmentIdentityKey(tableKey, key) {
  return JSON.stringify([tableKey, String(key)]);
}

function buildResolvedTable(table, key) {
  const pathSegments = table.pathSegments || normalizePathSegments(table.path);
  return {
    path: table.path,
    pathSegments,
    key,
    kind: "table",
  };
}

function buildResolvedAssignment(prefix, last, assignment, key) {
  const pathSegments = [...prefix, {
    value: String(last && last.value !== undefined ? last.value : ""),
    quoted: Boolean(last && last.quoted),
    range: last && last.range ? last.range : null,
  }];
  return {
    path: assignment.fullPath,
    pathSegments,
    key,
    kind: "assignment",
  };
}
